{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ebbettin/UCH_SRL/blob/main/Proteoform_analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Proteoform Analysis**"
      ],
      "metadata": {
        "id": "wzi2WW4Mb0_w"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Step 1: Install Required Libraries**\n",
        "\n",
        "This section installs the necessary libraries for the analysis: Biopython, pandas, openpyxl, MAFFT, and xlsxwriter."
      ],
      "metadata": {
        "id": "_nWCbUxAZc0E"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython pandas openpyxl"
      ],
      "metadata": {
        "id": "DTEOCkZ9jdPK",
        "outputId": "6a7c3e7e-9e70-46f0-c139-65cb936fb5a9",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (2.2.2)\n",
            "Requirement already satisfied: openpyxl in /usr/local/lib/python3.11/dist-packages (3.1.5)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: et-xmlfile in /usr/local/lib/python3.11/dist-packages (from openpyxl) (2.0.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m27.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.85\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install mafft"
      ],
      "metadata": {
        "id": "hWBMMYLEeHcM",
        "outputId": "d5794aa0-59c8-4ae8-d90c-9aa73c1268a1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  fonts-lato libauthen-sasl-perl libclone-perl libdata-dump-perl\n",
            "  libencode-locale-perl libfile-listing-perl libfont-afm-perl\n",
            "  libhtml-form-perl libhtml-format-perl libhtml-parser-perl\n",
            "  libhtml-tagset-perl libhtml-tree-perl libhttp-cookies-perl\n",
            "  libhttp-daemon-perl libhttp-date-perl libhttp-message-perl\n",
            "  libhttp-negotiate-perl libio-html-perl libio-socket-ssl-perl\n",
            "  liblwp-mediatypes-perl liblwp-protocol-https-perl libmailtools-perl\n",
            "  libnet-http-perl libnet-smtp-ssl-perl libnet-ssleay-perl libruby3.0\n",
            "  libtry-tiny-perl liburi-perl libwww-perl libwww-robotrules-perl lynx\n",
            "  lynx-common netbase perl-openssl-defaults rake ruby ruby-net-telnet\n",
            "  ruby-rubygems ruby-webrick ruby-xmlrpc ruby3.0 rubygems-integration\n",
            "Suggested packages:\n",
            "  libdigest-hmac-perl libgssapi-perl libcrypt-ssleay-perl libsub-name-perl\n",
            "  libbusiness-isbn-perl libauthen-ntlm-perl ri ruby-dev bundler\n",
            "Recommended packages:\n",
            "  blast2\n",
            "The following NEW packages will be installed:\n",
            "  fonts-lato libauthen-sasl-perl libclone-perl libdata-dump-perl\n",
            "  libencode-locale-perl libfile-listing-perl libfont-afm-perl\n",
            "  libhtml-form-perl libhtml-format-perl libhtml-parser-perl\n",
            "  libhtml-tagset-perl libhtml-tree-perl libhttp-cookies-perl\n",
            "  libhttp-daemon-perl libhttp-date-perl libhttp-message-perl\n",
            "  libhttp-negotiate-perl libio-html-perl libio-socket-ssl-perl\n",
            "  liblwp-mediatypes-perl liblwp-protocol-https-perl libmailtools-perl\n",
            "  libnet-http-perl libnet-smtp-ssl-perl libnet-ssleay-perl libruby3.0\n",
            "  libtry-tiny-perl liburi-perl libwww-perl libwww-robotrules-perl lynx\n",
            "  lynx-common mafft netbase perl-openssl-defaults rake ruby ruby-net-telnet\n",
            "  ruby-rubygems ruby-webrick ruby-xmlrpc ruby3.0 rubygems-integration\n",
            "0 upgraded, 43 newly installed, 0 to remove and 35 not upgraded.\n",
            "Need to get 12.3 MB of archives.\n",
            "After this operation, 51.2 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-lato all 2.0-2.1 [2,696 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 netbase all 6.3 [12.9 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/main amd64 libclone-perl amd64 0.45-1build3 [11.0 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/main amd64 libdata-dump-perl all 1.25-1 [25.9 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/main amd64 libencode-locale-perl all 1.05-1.1 [11.8 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhttp-date-perl all 6.05-1 [9,920 B]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/main amd64 libfile-listing-perl all 6.14-1 [11.2 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy/main amd64 libfont-afm-perl all 1.20-3 [13.6 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhtml-tagset-perl all 3.20-4 [12.5 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy/main amd64 liburi-perl all 5.10-1 [78.8 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhtml-parser-perl amd64 3.76-1build2 [88.4 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy/main amd64 libio-html-perl all 1.004-2 [15.4 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy/main amd64 liblwp-mediatypes-perl all 6.04-1 [19.5 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhttp-message-perl all 6.36-1 [76.8 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhtml-form-perl all 6.07-1 [22.2 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhtml-tree-perl all 5.07-2 [200 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhtml-format-perl all 2.12-1.1 [41.3 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhttp-cookies-perl all 6.10-1 [18.4 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libhttp-daemon-perl all 6.13-1ubuntu0.1 [22.9 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu jammy/main amd64 libhttp-negotiate-perl all 6.01-1 [12.5 kB]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu jammy/main amd64 perl-openssl-defaults amd64 5build2 [7,542 B]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu jammy/main amd64 libnet-ssleay-perl amd64 1.92-1build2 [327 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu jammy/main amd64 libio-socket-ssl-perl all 2.074-2 [192 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu jammy/main amd64 libnet-http-perl all 6.22-1 [23.2 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu jammy/main amd64 libtry-tiny-perl all 0.31-1 [21.8 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu jammy/main amd64 libwww-robotrules-perl all 6.02-1 [12.6 kB]\n",
            "Get:27 http://archive.ubuntu.com/ubuntu jammy/main amd64 libwww-perl all 6.61-1 [141 kB]\n",
            "Get:28 http://archive.ubuntu.com/ubuntu jammy/main amd64 liblwp-protocol-https-perl all 6.10-1 [10.9 kB]\n",
            "Get:29 http://archive.ubuntu.com/ubuntu jammy/main amd64 libnet-smtp-ssl-perl all 1.04-1 [5,948 B]\n",
            "Get:30 http://archive.ubuntu.com/ubuntu jammy/main amd64 libmailtools-perl all 2.21-1 [80.7 kB]\n",
            "Get:31 http://archive.ubuntu.com/ubuntu jammy/main amd64 rubygems-integration all 1.18 [5,336 B]\n",
            "Get:32 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 ruby3.0 amd64 3.0.2-7ubuntu2.10 [50.1 kB]\n",
            "Get:33 http://archive.ubuntu.com/ubuntu jammy/main amd64 ruby-rubygems all 3.3.5-2 [228 kB]\n",
            "Get:34 http://archive.ubuntu.com/ubuntu jammy/main amd64 ruby amd64 1:3.0~exp1 [5,100 B]\n",
            "Get:35 http://archive.ubuntu.com/ubuntu jammy/main amd64 rake all 13.0.6-2 [61.7 kB]\n",
            "Get:36 http://archive.ubuntu.com/ubuntu jammy/main amd64 ruby-net-telnet all 0.1.1-2 [12.6 kB]\n",
            "Get:37 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 ruby-webrick all 1.7.0-3ubuntu0.1 [52.1 kB]\n",
            "Get:38 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 ruby-xmlrpc all 0.3.2-1ubuntu0.1 [24.9 kB]\n",
            "Get:39 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libruby3.0 amd64 3.0.2-7ubuntu2.10 [5,114 kB]\n",
            "Get:40 http://archive.ubuntu.com/ubuntu jammy/universe amd64 lynx-common all 2.9.0dev.10-1 [1,024 kB]\n",
            "Get:41 http://archive.ubuntu.com/ubuntu jammy/universe amd64 mafft amd64 7.490-1 [776 kB]\n",
            "Get:42 http://archive.ubuntu.com/ubuntu jammy/main amd64 libauthen-sasl-perl all 2.1600-1.1 [43.1 kB]\n",
            "Get:43 http://archive.ubuntu.com/ubuntu jammy/universe amd64 lynx amd64 2.9.0dev.10-1 [719 kB]\n",
            "Fetched 12.3 MB in 1s (15.9 MB/s)\n",
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package fonts-lato.\n",
            "(Reading database ... 126281 files and directories currently installed.)\n",
            "Preparing to unpack .../00-fonts-lato_2.0-2.1_all.deb ...\n",
            "Unpacking fonts-lato (2.0-2.1) ...\n",
            "Selecting previously unselected package netbase.\n",
            "Preparing to unpack .../01-netbase_6.3_all.deb ...\n",
            "Unpacking netbase (6.3) ...\n",
            "Selecting previously unselected package libclone-perl.\n",
            "Preparing to unpack .../02-libclone-perl_0.45-1build3_amd64.deb ...\n",
            "Unpacking libclone-perl (0.45-1build3) ...\n",
            "Selecting previously unselected package libdata-dump-perl.\n",
            "Preparing to unpack .../03-libdata-dump-perl_1.25-1_all.deb ...\n",
            "Unpacking libdata-dump-perl (1.25-1) ...\n",
            "Selecting previously unselected package libencode-locale-perl.\n",
            "Preparing to unpack .../04-libencode-locale-perl_1.05-1.1_all.deb ...\n",
            "Unpacking libencode-locale-perl (1.05-1.1) ...\n",
            "Selecting previously unselected package libhttp-date-perl.\n",
            "Preparing to unpack .../05-libhttp-date-perl_6.05-1_all.deb ...\n",
            "Unpacking libhttp-date-perl (6.05-1) ...\n",
            "Selecting previously unselected package libfile-listing-perl.\n",
            "Preparing to unpack .../06-libfile-listing-perl_6.14-1_all.deb ...\n",
            "Unpacking libfile-listing-perl (6.14-1) ...\n",
            "Selecting previously unselected package libfont-afm-perl.\n",
            "Preparing to unpack .../07-libfont-afm-perl_1.20-3_all.deb ...\n",
            "Unpacking libfont-afm-perl (1.20-3) ...\n",
            "Selecting previously unselected package libhtml-tagset-perl.\n",
            "Preparing to unpack .../08-libhtml-tagset-perl_3.20-4_all.deb ...\n",
            "Unpacking libhtml-tagset-perl (3.20-4) ...\n",
            "Selecting previously unselected package liburi-perl.\n",
            "Preparing to unpack .../09-liburi-perl_5.10-1_all.deb ...\n",
            "Unpacking liburi-perl (5.10-1) ...\n",
            "Selecting previously unselected package libhtml-parser-perl:amd64.\n",
            "Preparing to unpack .../10-libhtml-parser-perl_3.76-1build2_amd64.deb ...\n",
            "Unpacking libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Selecting previously unselected package libio-html-perl.\n",
            "Preparing to unpack .../11-libio-html-perl_1.004-2_all.deb ...\n",
            "Unpacking libio-html-perl (1.004-2) ...\n",
            "Selecting previously unselected package liblwp-mediatypes-perl.\n",
            "Preparing to unpack .../12-liblwp-mediatypes-perl_6.04-1_all.deb ...\n",
            "Unpacking liblwp-mediatypes-perl (6.04-1) ...\n",
            "Selecting previously unselected package libhttp-message-perl.\n",
            "Preparing to unpack .../13-libhttp-message-perl_6.36-1_all.deb ...\n",
            "Unpacking libhttp-message-perl (6.36-1) ...\n",
            "Selecting previously unselected package libhtml-form-perl.\n",
            "Preparing to unpack .../14-libhtml-form-perl_6.07-1_all.deb ...\n",
            "Unpacking libhtml-form-perl (6.07-1) ...\n",
            "Selecting previously unselected package libhtml-tree-perl.\n",
            "Preparing to unpack .../15-libhtml-tree-perl_5.07-2_all.deb ...\n",
            "Unpacking libhtml-tree-perl (5.07-2) ...\n",
            "Selecting previously unselected package libhtml-format-perl.\n",
            "Preparing to unpack .../16-libhtml-format-perl_2.12-1.1_all.deb ...\n",
            "Unpacking libhtml-format-perl (2.12-1.1) ...\n",
            "Selecting previously unselected package libhttp-cookies-perl.\n",
            "Preparing to unpack .../17-libhttp-cookies-perl_6.10-1_all.deb ...\n",
            "Unpacking libhttp-cookies-perl (6.10-1) ...\n",
            "Selecting previously unselected package libhttp-daemon-perl.\n",
            "Preparing to unpack .../18-libhttp-daemon-perl_6.13-1ubuntu0.1_all.deb ...\n",
            "Unpacking libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libhttp-negotiate-perl.\n",
            "Preparing to unpack .../19-libhttp-negotiate-perl_6.01-1_all.deb ...\n",
            "Unpacking libhttp-negotiate-perl (6.01-1) ...\n",
            "Selecting previously unselected package perl-openssl-defaults:amd64.\n",
            "Preparing to unpack .../20-perl-openssl-defaults_5build2_amd64.deb ...\n",
            "Unpacking perl-openssl-defaults:amd64 (5build2) ...\n",
            "Selecting previously unselected package libnet-ssleay-perl:amd64.\n",
            "Preparing to unpack .../21-libnet-ssleay-perl_1.92-1build2_amd64.deb ...\n",
            "Unpacking libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Selecting previously unselected package libio-socket-ssl-perl.\n",
            "Preparing to unpack .../22-libio-socket-ssl-perl_2.074-2_all.deb ...\n",
            "Unpacking libio-socket-ssl-perl (2.074-2) ...\n",
            "Selecting previously unselected package libnet-http-perl.\n",
            "Preparing to unpack .../23-libnet-http-perl_6.22-1_all.deb ...\n",
            "Unpacking libnet-http-perl (6.22-1) ...\n",
            "Selecting previously unselected package libtry-tiny-perl.\n",
            "Preparing to unpack .../24-libtry-tiny-perl_0.31-1_all.deb ...\n",
            "Unpacking libtry-tiny-perl (0.31-1) ...\n",
            "Selecting previously unselected package libwww-robotrules-perl.\n",
            "Preparing to unpack .../25-libwww-robotrules-perl_6.02-1_all.deb ...\n",
            "Unpacking libwww-robotrules-perl (6.02-1) ...\n",
            "Selecting previously unselected package libwww-perl.\n",
            "Preparing to unpack .../26-libwww-perl_6.61-1_all.deb ...\n",
            "Unpacking libwww-perl (6.61-1) ...\n",
            "Selecting previously unselected package liblwp-protocol-https-perl.\n",
            "Preparing to unpack .../27-liblwp-protocol-https-perl_6.10-1_all.deb ...\n",
            "Unpacking liblwp-protocol-https-perl (6.10-1) ...\n",
            "Selecting previously unselected package libnet-smtp-ssl-perl.\n",
            "Preparing to unpack .../28-libnet-smtp-ssl-perl_1.04-1_all.deb ...\n",
            "Unpacking libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Selecting previously unselected package libmailtools-perl.\n",
            "Preparing to unpack .../29-libmailtools-perl_2.21-1_all.deb ...\n",
            "Unpacking libmailtools-perl (2.21-1) ...\n",
            "Selecting previously unselected package rubygems-integration.\n",
            "Preparing to unpack .../30-rubygems-integration_1.18_all.deb ...\n",
            "Unpacking rubygems-integration (1.18) ...\n",
            "Selecting previously unselected package ruby3.0.\n",
            "Preparing to unpack .../31-ruby3.0_3.0.2-7ubuntu2.10_amd64.deb ...\n",
            "Unpacking ruby3.0 (3.0.2-7ubuntu2.10) ...\n",
            "Selecting previously unselected package ruby-rubygems.\n",
            "Preparing to unpack .../32-ruby-rubygems_3.3.5-2_all.deb ...\n",
            "Unpacking ruby-rubygems (3.3.5-2) ...\n",
            "Selecting previously unselected package ruby.\n",
            "Preparing to unpack .../33-ruby_1%3a3.0~exp1_amd64.deb ...\n",
            "Unpacking ruby (1:3.0~exp1) ...\n",
            "Selecting previously unselected package rake.\n",
            "Preparing to unpack .../34-rake_13.0.6-2_all.deb ...\n",
            "Unpacking rake (13.0.6-2) ...\n",
            "Selecting previously unselected package ruby-net-telnet.\n",
            "Preparing to unpack .../35-ruby-net-telnet_0.1.1-2_all.deb ...\n",
            "Unpacking ruby-net-telnet (0.1.1-2) ...\n",
            "Selecting previously unselected package ruby-webrick.\n",
            "Preparing to unpack .../36-ruby-webrick_1.7.0-3ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Selecting previously unselected package ruby-xmlrpc.\n",
            "Preparing to unpack .../37-ruby-xmlrpc_0.3.2-1ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libruby3.0:amd64.\n",
            "Preparing to unpack .../38-libruby3.0_3.0.2-7ubuntu2.10_amd64.deb ...\n",
            "Unpacking libruby3.0:amd64 (3.0.2-7ubuntu2.10) ...\n",
            "Selecting previously unselected package lynx-common.\n",
            "Preparing to unpack .../39-lynx-common_2.9.0dev.10-1_all.deb ...\n",
            "Unpacking lynx-common (2.9.0dev.10-1) ...\n",
            "Selecting previously unselected package mafft.\n",
            "Preparing to unpack .../40-mafft_7.490-1_amd64.deb ...\n",
            "Unpacking mafft (7.490-1) ...\n",
            "Selecting previously unselected package libauthen-sasl-perl.\n",
            "Preparing to unpack .../41-libauthen-sasl-perl_2.1600-1.1_all.deb ...\n",
            "Unpacking libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Selecting previously unselected package lynx.\n",
            "Preparing to unpack .../42-lynx_2.9.0dev.10-1_amd64.deb ...\n",
            "Unpacking lynx (2.9.0dev.10-1) ...\n",
            "Setting up libhttp-date-perl (6.05-1) ...\n",
            "Setting up fonts-lato (2.0-2.1) ...\n",
            "Setting up libfile-listing-perl (6.14-1) ...\n",
            "Setting up libfont-afm-perl (1.20-3) ...\n",
            "Setting up mafft (7.490-1) ...\n",
            "Setting up libclone-perl (0.45-1build3) ...\n",
            "Setting up libhtml-tagset-perl (3.20-4) ...\n",
            "Setting up libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Setting up liblwp-mediatypes-perl (6.04-1) ...\n",
            "Setting up libtry-tiny-perl (0.31-1) ...\n",
            "Setting up perl-openssl-defaults:amd64 (5build2) ...\n",
            "Setting up libencode-locale-perl (1.05-1.1) ...\n",
            "Setting up rubygems-integration (1.18) ...\n",
            "Setting up libdata-dump-perl (1.25-1) ...\n",
            "Setting up ruby-net-telnet (0.1.1-2) ...\n",
            "Setting up libio-html-perl (1.004-2) ...\n",
            "Setting up lynx-common (2.9.0dev.10-1) ...\n",
            "Setting up ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Setting up netbase (6.3) ...\n",
            "Setting up lynx (2.9.0dev.10-1) ...\n",
            "update-alternatives: using /usr/bin/lynx to provide /usr/bin/www-browser (www-browser) in auto mode\n",
            "Setting up ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Setting up liburi-perl (5.10-1) ...\n",
            "Setting up libhttp-message-perl (6.36-1) ...\n",
            "Setting up libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Setting up libhttp-negotiate-perl (6.01-1) ...\n",
            "Setting up libhttp-cookies-perl (6.10-1) ...\n",
            "Setting up libnet-http-perl (6.22-1) ...\n",
            "Setting up libwww-robotrules-perl (6.02-1) ...\n",
            "Setting up libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Setting up libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Setting up libio-socket-ssl-perl (2.074-2) ...\n",
            "Setting up libhtml-form-perl (6.07-1) ...\n",
            "Setting up libhtml-tree-perl (5.07-2) ...\n",
            "Setting up libhtml-format-perl (2.12-1.1) ...\n",
            "Setting up libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Setting up libmailtools-perl (2.21-1) ...\n",
            "Setting up libwww-perl (6.61-1) ...\n",
            "Setting up ruby3.0 (3.0.2-7ubuntu2.10) ...\n",
            "Setting up liblwp-protocol-https-perl (6.10-1) ...\n",
            "Setting up ruby (1:3.0~exp1) ...\n",
            "Setting up rake (13.0.6-2) ...\n",
            "Setting up libruby3.0:amd64 (3.0.2-7ubuntu2.10) ...\n",
            "Setting up ruby-rubygems (3.3.5-2) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for mailcap (3.70+nmu1ubuntu1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install xlsxwriter"
      ],
      "metadata": {
        "id": "N8aHrN_l165F",
        "outputId": "efd49504-64d3-4bf4-d98e-15af5bf2e3f3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting xlsxwriter\n",
            "  Downloading xlsxwriter-3.2.5-py3-none-any.whl.metadata (2.7 kB)\n",
            "Downloading xlsxwriter-3.2.5-py3-none-any.whl (172 kB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/172.3 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[90m╺\u001b[0m\u001b[90m━\u001b[0m \u001b[32m163.8/172.3 kB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m172.3/172.3 kB\u001b[0m \u001b[31m3.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: xlsxwriter\n",
            "Successfully installed xlsxwriter-3.2.5\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Step 2: Sequence Alignment**\n",
        "\n",
        "This section will generate alignment files.\n",
        "Sequences with a stop codon before the last position will be excluded and their headers annotated in a txt file.\n",
        "\n",
        "Replace input_dir to reflect the folder containing your fasta files.\n",
        "Replace output_dir and output_txt_path to reflect where you want to save the output files."
      ],
      "metadata": {
        "id": "b07UZLXaZotS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "THLfv_-faNDY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio.Align.Applications import MafftCommandline\n",
        "\n",
        "# Replace input_dir and output_dir as required.\n",
        "input_dir = \"/content/seq\"\n",
        "output_dir = \"/content/aln\"\n",
        "\n",
        "# Create the output directory if it does not exist\n",
        "if not os.path.exists(output_dir):\n",
        "    os.makedirs(output_dir)\n",
        "\n",
        "# Define the path for the text file\n",
        "output_txt_path = \"/content/excluded_sequences.txt\"\n",
        "\n",
        "# DNA stop codons\n",
        "stop_codons_dna = {'TAA', 'TAG', 'TGA'}\n",
        "\n",
        "# Open the text file for writing\n",
        "with open(output_txt_path, \"w\") as output_txt_file:\n",
        "\n",
        "    # Iterate over all files in the input directory\n",
        "    for file in os.listdir(input_dir):\n",
        "        # Check if the file is a fasta file\n",
        "        if file.endswith(\".fasta\"):\n",
        "            input_path = os.path.join(input_dir, file)\n",
        "\n",
        "            # Read the sequences and filter out those with a stop codon before the last position\n",
        "            sequences = []\n",
        "            excluded_sequences = []\n",
        "            for record in SeqIO.parse(input_path, \"fasta\"):\n",
        "                sequence = record.seq\n",
        "\n",
        "                # Check for a stop codon before the last position\n",
        "                if any(sequence[i:i + 3] in stop_codons_dna for i in range(0, len(sequence) - 3, 3)):\n",
        "                    excluded_sequences.append(record.id)\n",
        "                else:\n",
        "                    sequences.append(record)\n",
        "\n",
        "            # Print the names of the excluded sequences to the console\n",
        "            if excluded_sequences:\n",
        "                print(f\"Excluded sequences with a stop codon before the last position in {file}: {', '.join(excluded_sequences)}\")\n",
        "\n",
        "                # Write the information to the text file\n",
        "                output_txt_file.write(f\"Excluded sequences in {file}: {', '.join(excluded_sequences)}\\n\")\n",
        "\n",
        "            # Count the number of sequences in the input file\n",
        "            num_sequences = len(sequences)\n",
        "\n",
        "            if num_sequences > 0:\n",
        "                # Create a temporary file for input to MAFFT\n",
        "                temp_input_path = \"/content/temp.fasta\"\n",
        "                with open(temp_input_path, \"w\") as temp_input_file:\n",
        "                    SeqIO.write(sequences, temp_input_file, \"fasta\")\n",
        "\n",
        "                # Run MAFFT with --auto and --quiet\n",
        "                output_filename = f\"{file.split('.')[0]}.fasta\"\n",
        "                output_path = os.path.join(output_dir, output_filename)\n",
        "                mafft_cline = MafftCommandline(input=temp_input_path, auto=True, quiet=True)\n",
        "                stdout, stderr = mafft_cline()\n",
        "\n",
        "                # Write the aligned sequences to the output file\n",
        "                with open(output_path, \"w\") as output_file:\n",
        "                    output_file.write(stdout)\n",
        "\n",
        "                # Print the output path and the number of sequences\n",
        "                print(f\"Alignment of {num_sequences} sequences from {file} was completed. Aligned sequences saved to {output_path}\")\n",
        "\n",
        "# Print the path to the text file\n",
        "print(f\"Excluded sequences information saved to: {output_txt_path}\")\n",
        "\n"
      ],
      "metadata": {
        "id": "CkM_yCmhVDAl",
        "outputId": "391dc3c6-dc04-454c-9d8e-ac6b32f0dd38",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/Bio/Application/__init__.py:39: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n",
            "\n",
            "Due to the on going maintenance burden of keeping command line application\n",
            "wrappers up to date, we have decided to deprecate and eventually remove these\n",
            "modules.\n",
            "\n",
            "We instead now recommend building your command line and invoking it directly\n",
            "with the subprocess module.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Alignment of 86 sequences from TP0515 combined.fasta was completed. Aligned sequences saved to /content/aln/TP0515 combined.fasta\n",
            "Excluded sequences information saved to: /content/excluded_sequences.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Step 3: Proteoform Analysis**\n",
        "\n",
        "This section will generate sheets containing information about the identified proteforms.\n",
        "\n",
        "Analysis will be made using the raw sequences files and the alignent previously generated."
      ],
      "metadata": {
        "id": "LLLep5HYa58q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import pandas as pd\n",
        "from openpyxl import Workbook, load_workbook\n",
        "from openpyxl.styles import PatternFill\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "\n",
        "# Directory path containing the input sequences\n",
        "sequence_directory = '/content/seq'\n",
        "\n",
        "# Directory path containing the output alignment files\n",
        "alignment_directory = '/content/aln'\n",
        "\n",
        "# Create the \"tables\" folder to store the Excel files\n",
        "output_folder = '/content/tables'\n",
        "os.makedirs(output_folder, exist_ok=True)\n",
        "\n",
        "# Create an empty list to store the Excel file paths\n",
        "excel_files = []\n",
        "\n",
        "# Define a function to split a sequence into codons\n",
        "def split_into_codons(sequence):\n",
        "    codons = [sequence[i:i+3] for i in range(0, len(sequence), 3)]\n",
        "    return codons\n",
        "\n",
        "# Define a function to get amino acid from codon\n",
        "def codon_to_aa(codon):\n",
        "    try:\n",
        "        translation = Seq(codon).translate()\n",
        "        return str(translation) if str(translation) != '-' else \"-\"\n",
        "    except Exception:\n",
        "        return \"-\"\n",
        "\n",
        "# Define a function to analyze sequences and save results to Excel\n",
        "def analyze_and_save_to_excel(file_path, sheet_name):\n",
        "\n",
        "    # Read the sequence file and extract the sequences\n",
        "    sequences_codons = {}\n",
        "\n",
        "    for record in SeqIO.parse(file_path, 'fasta'):\n",
        "        sequence = str(record.seq)\n",
        "        codons = split_into_codons(sequence)\n",
        "        sequences_codons[record.id] = codons\n",
        "\n",
        "    # Create a DataFrame from the sequences and their codons\n",
        "    df = pd.DataFrame.from_dict(sequences_codons, orient='index')\n",
        "\n",
        "    # Analyze each sequence individually to stop at its own stop codon\n",
        "    row_data_dict = {}\n",
        "    for strain, codons in df.iterrows():\n",
        "        row_data = []\n",
        "        for idx, codon in enumerate(codons):\n",
        "            aa = codon_to_aa(codon)\n",
        "            row_data.append((idx, f\"{codon} ({aa})\"))\n",
        "            if aa == '*':\n",
        "                break\n",
        "        row_data_dict[strain] = row_data\n",
        "\n",
        "    # Identify codon positions that occurred before stop codons\n",
        "    included_positions = set()\n",
        "    for row in row_data_dict.values():\n",
        "        included_positions.update(idx for idx, val in row)\n",
        "\n",
        "    # Filter columns that are in the included positions and are non-conserved\n",
        "    filtered_cols = [idx for idx in included_positions if df[idx].nunique() > 1]\n",
        "    filtered_cols = sorted(filtered_cols)\n",
        "\n",
        "    # Create an Excel workbook or load it if it already exists\n",
        "    output_file = os.path.join(output_folder, file_path.split('/')[-1].replace('.fasta', '_analysis.xlsx'))\n",
        "    if os.path.exists(output_file):\n",
        "        workbook = load_workbook(output_file)\n",
        "    else:\n",
        "        workbook = Workbook()\n",
        "        default_sheet = workbook.active\n",
        "        workbook.remove(default_sheet)\n",
        "\n",
        "    # Create a new sheet or load the existing one\n",
        "    if sheet_name in workbook.sheetnames:\n",
        "        sheet = workbook[sheet_name]\n",
        "    else:\n",
        "        sheet = workbook.create_sheet(title=sheet_name)\n",
        "\n",
        "    # Add header row\n",
        "    header_row = [\"\"] + [col + 1 for col in filtered_cols]\n",
        "    sheet.append(header_row)\n",
        "\n",
        "    # Write sequence rows\n",
        "    for strain, row_data in row_data_dict.items():\n",
        "        row_dict = dict(row_data)\n",
        "        row_out = [row_dict.get(col, \"\") for col in filtered_cols]\n",
        "        sheet.append([strain] + row_out)\n",
        "\n",
        "    # Get unique codons from the modified cells\n",
        "    unique_codons = set()\n",
        "    for row in sheet.iter_rows(min_row=2, min_col=2):\n",
        "        for cell in row:\n",
        "            if isinstance(cell.value, str) and cell.value:\n",
        "                codon = cell.value.split()[0]\n",
        "                unique_codons.add(codon)\n",
        "\n",
        "    # Generate consistent colors for unique codons\n",
        "    def generate_pastel_color():\n",
        "        pastel_range = range(150, 256)\n",
        "        color = 'FF' + ''.join(hex(c)[2:].rjust(2, '0') for c in [\n",
        "            os.urandom(1)[0] % len(pastel_range) + pastel_range.start for _ in range(3)\n",
        "        ])\n",
        "        return color\n",
        "\n",
        "    color_fills = {}\n",
        "    for value in unique_codons:\n",
        "        new_fill = PatternFill(\n",
        "            start_color=generate_pastel_color(),\n",
        "            end_color=generate_pastel_color(),\n",
        "            fill_type='solid'\n",
        "        )\n",
        "        color_fills[value] = new_fill\n",
        "\n",
        "    # Apply colors to cells\n",
        "    for row_num in range(2, sheet.max_row + 1):\n",
        "        for col_num in range(2, sheet.max_column + 1):\n",
        "            cell = sheet.cell(row=row_num, column=col_num)\n",
        "            value = cell.value\n",
        "            if isinstance(value, str) and value:\n",
        "                codon = value.split()[0]\n",
        "                if codon in color_fills:\n",
        "                    cell.fill = color_fills[codon]\n",
        "\n",
        "    # Save the workbook as an Excel file\n",
        "    workbook.save(output_file)\n",
        "    excel_files.append(output_file)\n",
        "\n",
        "# Process files in sequence and alignment directories\n",
        "for filename in os.listdir(sequence_directory):\n",
        "    if filename.endswith('.fasta'):\n",
        "        sequence_file_path = os.path.join(sequence_directory, filename)\n",
        "        analyze_and_save_to_excel(sequence_file_path, 'Original')\n",
        "\n",
        "for filename in os.listdir(alignment_directory):\n",
        "    if filename.endswith('.fasta') or filename.endswith('.aln'):\n",
        "        alignment_file_path = os.path.join(alignment_directory, filename)\n",
        "        analyze_and_save_to_excel(alignment_file_path, 'Alignment')\n",
        "\n",
        "# Output Excel file paths\n",
        "print(\"Excel files saved in the 'tables' folder:\")\n",
        "for file in excel_files:\n",
        "    print(file)"
      ],
      "metadata": {
        "id": "lNA6AxaD3NTm",
        "outputId": "9dbaea6c-326c-4bb6-ee74-6bc50a46f129",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Excel files saved in the 'tables' folder:\n",
            "/content/tables/TP0515 combined_analysis.xlsx\n",
            "/content/tables/TP0515 combined_analysis.xlsx\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Step 4: Proteoform List**\n",
        "\n",
        "This section will generate sheets containing the list of proteoforms identified."
      ],
      "metadata": {
        "id": "HXLaQtFKba5F"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def group_proteins_by_sequence(input_fasta_file, output_excel_file):\n",
        "    # Read the fasta file and create a dictionary to group proteins by sequence\n",
        "    sequence_dict = {}\n",
        "    for record in SeqIO.parse(input_fasta_file, \"fasta\"):\n",
        "        sequence = str(record.seq)\n",
        "        header = record.id\n",
        "        if sequence in sequence_dict:\n",
        "            sequence_dict[sequence].append(header)\n",
        "        else:\n",
        "            sequence_dict[sequence] = [header]\n",
        "\n",
        "    # Create a DataFrame from the dictionary\n",
        "    df = pd.DataFrame.from_dict(sequence_dict, orient='index').transpose()\n",
        "\n",
        "    # Create output folder if it does not exist\n",
        "    output_folder = os.path.dirname(output_excel_file)\n",
        "    os.makedirs(output_folder, exist_ok=True)\n",
        "\n",
        "    # Write the DataFrame to an Excel file without transposing\n",
        "    df.to_excel(output_excel_file, index=False)\n",
        "\n",
        "# Folder path containing all the input fasta files\n",
        "input_folder = \"/content/seq\"\n",
        "\n",
        "# Output folder path for the Excel files\n",
        "output_folder = \"/content/list\"\n",
        "\n",
        "# Loop through all files in the input folder\n",
        "for filename in os.listdir(input_folder):\n",
        "    if filename.endswith(\".fasta\"):\n",
        "        input_fasta_file = os.path.join(input_folder, filename)\n",
        "        output_excel_file = os.path.join(output_folder, filename.replace(\".fasta\", \".xlsx\"))\n",
        "\n",
        "        # Call the function to group proteins and create the Excel sheet\n",
        "        group_proteins_by_sequence(input_fasta_file, output_excel_file)"
      ],
      "metadata": {
        "id": "7jBvCVtl1Lgs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Step 5: Export files**\n",
        "\n",
        "This section will export all the output files as .zip"
      ],
      "metadata": {
        "id": "8rLGdYZLbcXV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import shutil\n",
        "from google.colab import files\n",
        "\n",
        "shutil.make_archive('/content/aln', 'zip', '/content/aln')\n",
        "files.download('/content/aln.zip')\n",
        "\n",
        "shutil.make_archive('/content/tables', 'zip', '/content/tables')\n",
        "files.download('/content/tables.zip')\n",
        "\n",
        "shutil.make_archive('/content/list', 'zip', '/content/list')\n",
        "files.download('/content/list.zip')"
      ],
      "metadata": {
        "id": "4ATrgIHLnUi1"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}